src/core/observables/PidObservable.hpp

```diff
@@ -49,5 +53,42 @@ class PidObservable : virtual public Observable {
   std::vector<int> const &ids() const { return m_ids; }
 };
 
-} // Namespace Observables
+namespace detail {
+template <class T> struct shape_impl;
+
+template <> struct shape_impl<double> {
+  static std::vector<size_t> eval(size_t /* n_part */) { return {1}; }
+};
+template <class _, size_t N> struct shape_impl<Utils::Vector<_, N>> {
+  static std::vector<size_t> eval(size_t /* n_part */) { return {N}; }
+};
+template <class T> struct shape_impl<std::vector<T>> {
+  static std::vector<size_t> eval(size_t n_part) {
+    std::vector<size_t> ret{n_part};
+    boost::copy(shape_impl<T>::eval(n_part), std::back_inserter(ret));
+
+    return ret;
+  }
+};
+} // namespace detail
+
+template <class ObsType> class ParticleObservable : public PidObservable {
+public:
+  using PidObservable::PidObservable;
+  std::vector<size_t> shape() const override {
+    using std::declval;
+
+    return detail::shape_impl<decltype(declval<ObsType>()(
+        declval<Utils::Span<const Particle *const>>()))>::eval(ids().size());
+  }
+
+  std::vector<double>
+  evaluate(Utils::Span<const Particle *const> particles) const override {
```

Comment:
I think you should change the signature to taking a span of `std::reference_wrapper`. `reference_wrapper` is sometime a bit awkward to work with, but I think here it fits the bill, and it will make the downstream code clearer. With this you can push the null check up to the caller...